pkg/daemon/daemon.go

```diff
@@ -67,59 +71,112 @@ const (
 	pathDevNull = "/dev/null"
 )
 
-// New sets up the systemd and kubernetes connections needed to update the
-// machine.
+// New sets up the bare minimum local connections required to create a Daemon instance
 func New(
 	rootMount string,
 	nodeName string,
 	operatingSystem string,
 	nodeUpdaterClient NodeUpdaterClient,
-	client mcfgclientset.Interface,
-	kubeClient kubernetes.Interface,
 	fileSystemClient FileSystemClient,
-	nodeInformer coreinformersv1.NodeInformer,
+	onceFrom string,
 ) (*Daemon, error) {
 	loginClient, err := login1.New()
 	if err != nil {
 		return nil, fmt.Errorf("Error establishing connection to logind dbus: %v", err)
 	}
 
-	if err = loadNodeAnnotations(kubeClient.CoreV1().Nodes(), nodeName); err != nil {
-		return nil, err
-	}
-
 	osImageURL, osVersion, err := nodeUpdaterClient.GetBootedOSImageURL(rootMount)
 	if err != nil {
 		return nil, fmt.Errorf("Error reading osImageURL from rpm-ostree: %v", err)
 	}
+
 	glog.Infof("Booted osImageURL: %s (%s)", osImageURL, osVersion)
 
 	dn := &Daemon{
 		name:              nodeName,
 		OperatingSystem:   operatingSystem,
 		NodeUpdaterClient: nodeUpdaterClient,
 		loginClient:       loginClient,
-		client:            client,
-		kubeClient:        kubeClient,
 		rootMount:         rootMount,
 		fileSystemClient:  fileSystemClient,
 		bootedOSImageURL:  osImageURL,
+		onceFrom:          onceFrom,
 	}
 
+	return dn, nil
+}
+
+// NewClusterDrivenDaemon sets up the systemd and kubernetes connections needed to update the
+// machine.
+func NewClusterDrivenDaemon(
+	rootMount string,
+	nodeName string,
+	operatingSystem string,
+	nodeUpdaterClient NodeUpdaterClient,
+	kubeconfig string,
+	fileSystemClient FileSystemClient,
+	onceFrom string,
+	stopCh chan (struct{}),
+	componentName string,
+) (*Daemon, error) {
+	dn, err := New(
+		rootMount,
+		nodeName,
+		operatingSystem,
+		nodeUpdaterClient,
+		fileSystemClient,
+		onceFrom)
+
+	if err != nil {
+		return nil, err
+	}
+
+	cb, err := common.NewClientBuilder(kubeconfig)
+	if err != nil {
+		glog.Fatalf("error creating clients: %v", err)
+	}
+
+	dn.kubeClient = cb.KubeClientOrDie(componentName)
+	dn.client = cb.MachineConfigClientOrDie(componentName)
+
+	if err = loadNodeAnnotations(dn.kubeClient.CoreV1().Nodes(), nodeName); err != nil {
+		return nil, err
+	}
+	return dn, nil
+}
+
+// StartInformer initializes and starts the informers.
+func (dn *Daemon) StartInformer(stopCh chan (struct{}), nodeName, componentName, kubeconfig string) error {
+	cb, err := common.NewClientBuilder(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	ctx := common.CreateControllerContext(cb, stopCh, componentName)
+
+	ctx.KubeInformerFactory.Start(ctx.Stop)
```

Comment:
In the current workflow we do this after setting up the nodeInformer/Lister. I'm not sure if the ordering will cause issues, but I remember that is what we did elsewhere for MCO/MCC, so perhaps move this line and the following line to the end of this function?

Reply:
Updated and moved start to the end of this function before return.